### ISSUE TITLE
Formatting failed by vscode-cpptools/clangformat with no output

### ISSUE THREAD

**Issue by francoccc** [microsoft/vscode-cpptools#11643]
### Environment

- OS and Version: Mac OS 13.1 (22C65)
- VS Code Version: 1.84.0 (Universal)
- C/C++ Extension Version:  1.18.2-linux-x64
- If using SSH remote, specify OS of remote machine: Linux XxxBlah 4.19.0-16-amd64 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux

### Bug Summary and Steps to Reproduce

Bug Summary:
  I have set embedded clangFormat as my default cpp formatter. When I formatted the document (*.cpp), I had seen some errors. It hints to me to see the output window for details. However, I didn't find any error messages in the OUTPUT of C/C++. I have also run the same command in the terminal, but unfortunately, it was stuck.  By the way, I also tried on another machine which is the RHEL, and was affected by the same problem.

Steps to reproduce:
1. Format document
2. See error

Expected behavior:
  Formatting correctly.

### Configuration and Logs

```shell
c_cpp_properties.json (part of format):
...
  "editor.detectIndentation": false,
  "editor.formatOnSave": true,
  "editor.formatOnPaste": false,
  "editor.formatOnType": true,
  "editor.indentSize": "tabSize",
  "editor.tabSize": 2,
  "C_Cpp.autocompleteAddParentheses": true,
  "C_Cpp.formatting": "clangFormat", // In this project, we choose clangFormat as default
  "C_Cpp.vcFormat.indent.accessSpecifiers": false,
  "C_Cpp.vcFormat.indent.caseLabels": true,
  "C_Cpp.vcFormat.indent.caseContents": true,
  "C_Cpp.vcFormat.indent.namespaceContents": false,
  "C_Cpp.vcFormat.newLine.beforeOpenBrace.block": "sameLine",
  "C_Cpp.vcFormat.newLine.beforeOpenBrace.function": "sameLine",
  "C_Cpp.vcFormat.newLine.beforeOpenBrace.namespace": "sameLine",
  "C_Cpp.vcFormat.newLine.beforeOpenBrace.lambda": "sameLine",
  "C_Cpp.vcFormat.newLine.beforeOpenBrace.type": "sameLine",
  "C_Cpp.vcFormat.newLine.closeBraceSameLine.emptyFunction": true,
  "C_Cpp.vcFormat.newLine.closeBraceSameLine.emptyType": true,
  "C_Cpp.vcFormat.space.pointerReferenceAlignment": "right",
  "C_Cpp.vcFormat.space.beforeComma": false,
  "C_Cpp.vcFormat.space.afterComma": true,
  // The part of Clang format
  "C_Cpp.clang_format_sortIncludes": true,
  "C_Cpp.clang_format_fallbackStyle": "Google",
...

The OUTPUT of C/C++

Formatting failed:
$(home)/.vscode-server/extensions/ms-vscode.cpptools-1.18.2-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=LLVM -sort-includes=1 --Wno-error=unknown -offset=399 -length=901 -assume-filename=XXFile.cpp
Formatting failed:
$(home)/.vscode-server/extensions/ms-vscode.cpptools-1.18.2-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=LLVM -sort-includes=1 --Wno-error=unknown -assume-filename=XXFile.cpp
Formatting failed:
$(home)/.vscode-server/extensions/ms-vscode.cpptools-1.18.2-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=google -sort-includes=1 --Wno-error=unknown -offset=375 -length=925 -assume-filename=XXFile.cpp

.clang-format (Workspace):
---
# BasedOnStyle:  Google
Language: Cpp
AccessModifierOffset: -2
AlignEscapedNewlinesLeft: true
AlignTrailingComments: true
AllowAllParametersOfDeclarationOnNextLine: true
AllowShortIfStatementsOnASingleLine: false
AllowShortLoopsOnASingleLine: false
AlwaysBreakBeforeMultilineStrings: true
AlwaysBreakTemplateDeclarations: true
BinPackParameters: true
BreakBeforeBinaryOperators: true
BreakBeforeBraces: Attach
BreakBeforeTernaryOperators: true
BreakConstructorInitializersBeforeComma: false
BreakStringLiterals: true
ColumnLimit:     80
ConstructorInitializerAllOnOneLineOrOnePerLine: true
ConstructorInitializerIndentWidth: 4
ContinuationIndentWidth: 4
Cpp11BracedListStyle: true
DerivePointerBinding: true
ExperimentalAutoDetectBinPacking: false
IndentCaseLabels: true
IndentFunctionDeclarationAfterType: true
IndentWidth:     2
MaxEmptyLinesToKeep: 1
NamespaceIndentation: None
ObjCSpaceBeforeProtocolList: false
PenaltyBreakBeforeFirstCallParameter: 1
PenaltyBreakComment: 60
PenaltyBreakFirstLessLess: 120
PenaltyBreakString: 1000
PenaltyExcessCharacter: 1000000
PenaltyReturnTypeOnItsOwnLine: 200
PointerBindsToType: true
SpaceAfterControlStatementKeyword: true
SpaceBeforeAssignmentOperators: true
SpaceInEmptyParentheses: false
SpacesBeforeTrailingComments: 2
SpacesInAngles:  false
SpacesInCStyleCastParentheses: false
SpacesInParentheses: false
Standard:        Cpp11
TabWidth:        8
UseTab:          Never
...


ldd clang-format:
  statically linked
```

**Comment by sean-mcmanus** [comment]
@francoccc Does selecting all the code in the document and then formatting work? Does downgrading to 1.18.1 work?

I'm not able to repro the bug so far, so there might be some special condition that is causing it.

**Comment by francoccc** [comment]
@sean-mcmanus Thanks for your reply, I also think there might be some special condition that especially happened on working with remote servers.
 I have selected all the code in the document or the part of it, and then formatting work. I also tried to format the document without selection. After different tries, the error of formatting happened always.

Before I submitted this issue, I solved a duplicate key problem in `.clang-format` that is similar to #1259. It's a defect of clang-format and has no error message in the C/C++ OUTPUT. 
Since I have several developing servers, the solution of a duplicate key seems to be good for me. While one of the servers still has the problem of formatting failed. I did a lot including changing the C/C++ Extension Version. Finally it works well after I start a vscode client on Windows to connect the same.
I will continue to observe this problem. Maybe vcFormat is better than clang-format but it's difficult for me to change back.

**Comment by sean-mcmanus** [comment]
@francoccc If you run the command shows in the C/C++ logging window but remove the "-assume-filename" string, e.g.

```
$(home)/.vscode-server/extensions/ms-vscode.cpptools-1.18.2-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=LLVM -sort-includes=1 --Wno-error=unknown -offset=399 -length=901 XXFile.cpp
```

then you should be able to get more error info from clang-format.

**Comment by tehKaiN** [comment]
I have similar problem on win10 x64, 1.8.3. Output window shows:

```
Formatting failed:
c:\Users\tehKa\.vscode\extensions\ms-vscode.cpptools-1.18.3-win32-x64/bin/../LLVM/bin/clang-format.exe -style=file -fallback-style=LLVM --Wno-error=unknown -assume-filename=C:\prg\_hk\firmware2\modules\stdio\CLI.cpp
```

running clangformat from external terminal with that cmdline hangs indefinitely. Even running clangformat itself hangs until I ctrl+C it:

![obraz](https://github.com/microsoft/vscode-cpptools/assets/16180344/fde5bfbb-2e73-4f64-b45b-5f0daef13fab)

**Comment by opokatech** [comment]
I have identical problem - running on Linux, with this VS code 1.84.2-1699528352 with C/C++ plugin 1.18.4.

Actually the formatting worked well till **1.17.4** (incl.). Then, since 1.17.5 it does not. So I downgraded and since then I occasionally do upgrade and check if it works again. Today I checked against 1.18.4 with no luck.

Running the command that is visible on the C++ console `/home/mo/.vscode/extensions/ms-vscode.cpptools-1.18.4-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=llvm --Wno-error=unknown` just blocks it untill CTRL-C.

But this is the same for the working version.

 Is there a way to get more insight what is failing?

**Comment by Wzzzx** [comment]
> I have similar problem on win10 x64, 1.8.3. Output window shows:
> 
> ```
> Formatting failed:
> c:\Users\tehKa\.vscode\extensions\ms-vscode.cpptools-1.18.3-win32-x64/bin/../LLVM/bin/clang-format.exe -style=file -fallback-style=LLVM --Wno-error=unknown -assume-filename=C:\prg\_hk\firmware2\modules\stdio\CLI.cpp
> ```
> 
> running clangformat from external terminal with that cmdline hangs indefinitely. Even running clangformat itself hangs until I ctrl+C it:
> 
> ![obraz](https://user-images.githubusercontent.com/16180344/282769103-fde5bfbb-2e73-4f64-b45b-5f0daef13fab.png)

I have meet the same problem, and I downgrade to v1.17.5 to solve it.

**Comment by Nowaterisenough** [comment]
I downgraded to v1.17.4, but the problem still hasn't been resolved.

**Comment by opokatech** [comment]
For me v1.17.4 works and versions above it - don't.
I'm using own .clang-format file and clang-format is installed on my machine.

Is it possible to see some more debug when the formatting fails?

**Comment by sean-mcmanus** [comment]
@Wzzzx To get the failure reason, can you run the command: 

c:\Users\tehKa\.vscode\extensions\ms-vscode.cpptools-1.18.3-win32-x64/bin/../LLVM/bin/clang-format.exe -style=file -fallback-style=LLVM --Wno-error=unknown C:\prg\_hk\firmware2\modules\stdio\CLI.cpp

i.e. with the `-assume-filename=` removed? The command has to end with the filename or it'll get stuck waiting for input.

**Comment by opokatech** [comment]
OK - the problem was solved for me.

For debugging I took the line reported by the VScode and on the console used `cat file_to_be_formatted | the_script_call_with_all_parameters`. Then it was pretty clear what the problem was. I fixed it in .clang-format and the newest version of the plugin works like a charm.

**Comment by opokatech** [comment]
It would be **very, very helpful** to intercept the stderr output of the command and showing that along the message that "Formatting failed".

**Comment by bobbrow** [comment]
> It would be **very, very helpful** to intercept the stderr output of the command and show that along the message that "Formatting failed".

Hmm, I thought I did that a long time ago but maybe we only printed the stdout to the Output Window. https://github.com/Microsoft/vscode-cpptools/issues/1259#issuecomment-371972526

**Comment by sean-mcmanus** [comment]
@opokatech @bobbrow stderr clang-format logging should be available for 1.19.0.

**Comment by sean-mcmanus** [comment]
The logging error is fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.19.0 (pre-release).

It sounds like the duplicate keys in .clang-format from the clang-format 17 update might be the most common cause.

**Comment by opokatech** [comment]
@sean-mcmanus Thank you for the info! Indeed in my case it was about the keys duplication in the .clang-format.

**Comment by francoccc** [comment]
> @francoccc If you run the command shows in the C/C++ logging window but remove the "-assume-filename" string, e.g.
> 
> ```
> $(home)/.vscode-server/extensions/ms-vscode.cpptools-1.18.2-linux-x64/bin/../LLVM/bin/clang-format -style=file -fallback-style=LLVM -sort-includes=1 --Wno-error=unknown -offset=399 -length=901 XXFile.cpp
> ```
> 
> then you should be able to get more error info from clang-format.

Thanks.

**Comment by tehKaiN** [comment]
> stderr clang-format logging should be available for 1.19.0: #8944

yep, it made the problem's cause more apparent to me. There was a duplicate entry in my clangformat file. Thanks for that addition!
